calyx/src/analysis/dominator_map.rs

```diff
@@ -0,0 +1,489 @@
+            }
+            ir::Control::If(_) => {
+                hs.insert(Self::get_guaranteed_attribute(c, END_ID));
+            }
+            ir::Control::Seq(ir::Seq { stmts, .. }) => {
+                match (&stmts[..]).last() {
+                    None => panic!("error: empty Seq block. Run ___ "),
+                    Some(control) => return Self::get_final(control),
+                }
+            }
+            ir::Control::Par(ir::Par { stmts, .. }) => {
+                for stmt in stmts {
+                    let stmt_final = Self::get_final(stmt);
+                    hs = hs.union(&stmt_final).copied().collect()
+                }
+            }
+        }
+        hs
+    }
+
+    //Given an id and its predecessors pred, and a domination map d_map, updates
+    //d_map accordingly (i.e. the union of all dominators of the predecessors
+    //plus itself).
+    fn update_node(pred: &HashSet<u64>, id: u64, d_map: &mut DominatorMap) {
+        let mut union: HashSet<u64> = HashSet::new();
+        for id in pred.iter() {
+            if let Some(dominators) = d_map.map.get(id) {
+                union = union.union(dominators).copied().collect();
+            }
+        }
+        union.insert(id);
+        d_map.map.insert(id, union);
+    }
+
+    //Looks through each "node" in the "graph" and updates the dominators accordingly
+    fn update_map(
+        main_c: &ir::Control,
+        cur_id: u64,
+        pred: &HashSet<u64>,
+        d_map: &mut DominatorMap,
+    ) {
+        let c = match Self::get_control(cur_id, main_c) {
+            Some(control) => control,
+            None => return,
+        };
+        match c {
+            ir::Control::Empty(_) => {
+                unreachable!(
+                    "should not pattern match agaisnt empty in update_map()"
+                )
+            }
+            ir::Control::Invoke(_) | ir::Control::Enable(_) => {
+                Self::update_node(pred, cur_id, d_map);
+            }
+            ir::Control::Seq(ir::Seq { stmts, .. }) => {
+                //Could try to think a way of doing it w/o this first stuff
+                let mut first = true;
+                let mut prev_id = cur_id;
+                for stmt in stmts {
+                    let id = Self::get_id(stmt, true);
+                    if first {
+                        Self::update_map(main_c, id, pred, d_map);
+                        first = false;
+                    } else {
+                        Self::update_map(
+                            main_c,
+                            id,
+                            &d_map
+                                .exits_map
+                                .get(&prev_id)
+                                .unwrap_or_else(|| {
+                                    unreachable!(
+                                        "exit node map does not have value for {}",
+                                        prev_id
```

Comment:
```suggestion
                                        "exit node map does not have value for {prev_id}",
```

Reply:
When I try to do this locally, it gives me the following warning: 

panic message contains an unused formatting placeholder
`#[warn(non_fmt_panics)]` on by default

Reply:
if prefers: `unreachable!("{}", "exit node map does not have value for {prev_id}", )`